### PR TITLE
2950053 - Adding BotauthSubscriber to support SSO

### PR DESCRIPTION
**Pull request by elopezanaya**
## Description

The subscriber component is bound to the incoming activity from ACS , in this way it detects when a specific incoming event is present so it can act according to the use cases specified.

## Solution 

The subscriber will be actively checking over any incoming message from ACS, in this way it can catch and avoid the transmission of such message if indicated by the BotAuthConfig defined by the customer.

## Acceptance criteria 
- extract the SAS URL from the signIn Adaptive Card and pass it to the BotAuthTokenProvider function
- Communication to the scripting layer will be via BroadCastService
- It will communicate the presence of the adaptive card via the SigninCardReceived event
- it Will trigger the execution of the token provider via executeSigninCardCallbackRequest
- It will listen to the result of the execution via an event: executeSigninCardCallbackResponse
- it should honor the config defined by the user and block the pass of the activity when indicated or let it go as indicated.
- It needs to be resilient, in case of error or the missing result of execution will allow to pass the activity for this to be shown.
- It needs to support Popout

### Normal window, allow the card to be displayed

image

### Normal window, do not allow the card to be shown
image

### Pop out, allow card to be displayed

image

### Pop out , do not allow card to be displayed
image

### popout , no function defined , it allows card

image

### CODE REVIEW

File: chat-widget/src/common/telemetry/TelemetryConstants.ts [hunk 1]
```diff
@@ -48,7 +48,10 @@ export enum BroadcastEvent {
     ClosePopoutWindow = "ClosePopoutWindow",
     RaiseErrorEvent = "RaiseErrorEvent",
     NetworkDisconnected = "NetworkDisconnected",
-    NetworkReconnected = "NetworkReconnected"
+    NetworkReconnected = "NetworkReconnected",
+    SigninCardReceived = "lcw:signInCardReceived",
```
Comment (charliewang95) [inline chat-widget/src/common/telemetry/TelemetryConstants.ts:52]: Let’s just use “SigninCardReceived” and not start with “lcw”, Asian parity with other events

File: chat-widget/src/common/telemetry/TelemetryConstants.ts [hunk 2]
```diff
@@ -48,7 +48,10 @@ export enum BroadcastEvent {
...
+    ExecuteSigninCardCallbackRequest = "executeSigninCardCallbackRequest",
```
Comment (charliewang95) [inline chat-widget/src/common/telemetry/TelemetryConstants.ts:53]: Please start these 2 with Uppercase

File: chat-widget/src/common/telemetry/TelemetryConstants.ts [hunk 3]
```diff
@@ -141,6 +144,11 @@ export enum TelemetryEvent {
     CustomerVoiceResponsePageLoaded = "CustomerVoiceResponsePageLoaded",
     CustomerVoiceFormResponseSubmitted = "CustomerVoiceFormResponseSubmitted",
     CustomerVoiceFormResponseError = "CustomerVoiceFormResponseError",
+    BotAuthActivityEmptySasUrl = "BotAuthActivityEmptySasUrl",
```
Comment (charliewang95) [inline chat-widget/src/common/telemetry/TelemetryConstants.ts:147]: Please update Telemetry.md with all newly added TelemetryEvent and BroadcastEvent (the one we raise only. Don’t need to log the one we subscribe to, I.e. ExecuteSigninCardCallbackResponse)

File: chat-widget/src/common/telemetry/TelemetryConstants.ts (hunk 1 above)
Comment (charliewang95) [inline chat-widget/src/common/telemetry/TelemetryConstants.ts:52]: Please update ChangeLog.md

File: chat-widget/src/components/livechatwidget/common/ActivitySubscriber/BotAuthActivitySubscriber.ts [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+};
+
+const fetchBotAuthConfig = async (retries: number): Promise<any> => {
+    TelemetryHelper.logLoadingEvent(LogLevel.INFO, {
+        Event: TelemetryEvent.SetBotAuthProviderFetchConfig,
+    });
+
+    const executeSigningCardRequest: ICustomEvent = { eventName: BroadcastEvent.ExecuteSigninCardCallbackRequest };
+    BroadcastService.postMessage(executeSigningCardRequest);
+    const listener = BroadcastService.getMessageByEventName(BroadcastEvent.ExecuteSigninCardCallbackResponse)
```
Comment (charliewang95) [inline chat-widget/src/components/livechatwidget/common/ActivitySubscriber/BotAuthActivitySubscriber.ts:52]: Please add a separate documentation instructing consumers to use this feature. Refer to the new doc in the 2 README’s “See more” section at the bottom. The 2 readmes are under root folder and chat-widget folder

File: chat-widget/src/common/telemetry/TelemetryConstants.ts (hunk 1 above)
Comment (elopezanaya) [reply chat-widget/src/common/telemetry/TelemetryConstants.ts:52]: done

File: chat-widget/src/common/telemetry/TelemetryConstants.ts (hunk 2 above)
Comment (elopezanaya) [reply chat-widget/src/common/telemetry/TelemetryConstants.ts:53]: done

File: chat-widget/src/components/livechatwidget/common/ActivitySubscriber/BotAuthActivitySubscriber.ts [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+export class BotAuthActivitySubscriber implements IActivitySubscriber {
+    public observer: any;
+    private signInCardSeen: Set<string>;
+
+    constructor() {
+        this.signInCardSeen = new Set();
+    }
+
+    public applicable(activity: any): boolean {
+        return activity && activity.attachments && activity.attachments.length && activity.attachments[0] && supportedSignInCardContentTypes.indexOf(activity.attachments[0].contentType) >= 0;
```
[the remark this answers is not in the document]
Comment (elopezanaya) [reply chat-widget/src/components/livechatwidget/common/ActivitySubscriber/BotAuthActivitySubscriber.ts:79]: done

File: chat-widget/src/common/telemetry/TelemetryConstants.ts (hunk 3 above)
Comment (elopezanaya) [reply chat-widget/src/common/telemetry/TelemetryConstants.ts:147]: done

File: chat-widget/src/components/livechatwidget/common/ActivitySubscriber/BotAuthActivitySubscriber.ts [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    }
+    return sasUrl;
+};
+
+const fetchBotAuthConfig = async (retries: number): Promise<any> => {
+    TelemetryHelper.logLoadingEvent(LogLevel.INFO, {
+        Event: TelemetryEvent.SetBotAuthProviderFetchConfig,
+    });
+
+    const executeSigningCardRequest: ICustomEvent = { eventName: BroadcastEvent.ExecuteSigninCardCallbackRequest };
```
[the remark this answers is not in the document]
Comment (elopezanaya) [reply chat-widget/src/components/livechatwidget/common/ActivitySubscriber/BotAuthActivitySubscriber.ts:50]: done

File: chat-widget/src/components/livechatwidget/common/ActivitySubscriber/BotAuthActivitySubscriber.ts (hunk 4 above)
Comment (elopezanaya) [reply chat-widget/src/components/livechatwidget/common/ActivitySubscriber/BotAuthActivitySubscriber.ts:52]: wip

**Review by abhinab22mohanty — APPROVED** [review]

File: docs/Telemetry.md [hunk 7]
```diff
@@ -215,6 +215,10 @@ Refer to the below table to understand different critical telemetry events raise
 | `ReceivedNullOrEmptyToken` | On receiving null or empty auth token |
 | `SystemMessageReceived` | On system message received |
 | `HistoryMessageReceived` | On history message received |
+| `ExecuteSigninCardCallbackRequest` | On Signin adaptive card received from bot |
```
Comment (xTEddie) [inline docs/Telemetry.md:218]: Would `sign-in card` be better?

That's how the documentation call it:

https://learn.microsoft.com/en-us/microsoftteams/platform/task-modules-and-cards/cards/cards-reference#sign-in-card

File: chat-widget/src/common/telemetry/TelemetryConstants.ts [hunk 8]
```diff
@@ -48,7 +48,10 @@ export enum BroadcastEvent {
     ClosePopoutWindow = "ClosePopoutWindow",
     RaiseErrorEvent = "RaiseErrorEvent",
     NetworkDisconnected = "NetworkDisconnected",
-    NetworkReconnected = "NetworkReconnected"
+    NetworkReconnected = "NetworkReconnected",
+    SigninCardReceived = "SignInCardReceived",
+    ExecuteSigninCardCallbackRequest = "ExecuteSigninCardCallbackRequest",
```
Comment (xTEddie) [inline chat-widget/src/common/telemetry/TelemetryConstants.ts:53]: `SignInCardConfigRequest` or `BotAuthConfigRequest`?

Comment (elopezanaya) [reply chat-widget/src/common/telemetry/TelemetryConstants.ts:53]: done

File: docs/Telemetry.md (hunk 7 above)
Comment (elopezanaya) [reply docs/Telemetry.md:218]: done

File: docs/Telemetry.md [hunk 9]
```diff
@@ -215,6 +215,10 @@ Refer to the below table to understand different critical telemetry events raise
 | `ReceivedNullOrEmptyToken` | On receiving null or empty auth token |
 | `SystemMessageReceived` | On system message received |
 | `HistoryMessageReceived` | On history message received |
+| `SigninCardReceived` | On sign-in adaptive card received from bot |
+| `BotAuthActivityEmptySasUrl` | On detecting a signing adaptive card with no SAS Url |
+| `ExecuteSigninCardCallbackRequest` | On triger the execution of the BotAuthTokenProvider function |
```
Comment (xTEddie) [inline docs/Telemetry.md:220]: Need to update the Event naming

File: docs/Features.md [hunk 10]
```diff
@@ -133,4 +133,39 @@ ReactDOM.render(
     <LiveChatWidget {...liveChatWidgetProps}/>,
     document.getElementById("my-container")
 );
-```
\ No newline at end of file
+```
+
+### __Use of BotAuthConfigRequest and BotAuthConfigResponse for PVA SSO__
+
+```BotAuthActivitySubscriber``` monitors incoming activity from ACS to detect when a sign-in card is being passed from PVA to the customer asking for sign-in.
```
Comment (charliewang95) [inline docs/Features.md:140]: "activity" -> "activities"

File: docs/Features.md [hunk 11]
```diff
@@ -133,4 +133,39 @@ ReactDOM.render(
...
+                payload: {
+                    response: res,
+                    functionIsSet: BotAuthValues.botAuthTokenProvider?.functionIsSet
+                }
+            });
+        });
+```
+
+
+To setup ```setBotAuthTokenProvider``` at the scripting layer and set the function in the html , please follow up next documentation [setBotAuthTokenProvider guideline](https://learn.microsoft.com/en-us/dynamics365/customer-service/developer/reference/methods/setbotauthtokenprovider)
```
Comment (charliewang95) [inline docs/Features.md:171]: extra space after "html"

Comment (charliewang95) [inline docs/Features.md:171]: "please follow up next documentation" -> "please follow this documentation"

**Review by charliewang95 — APPROVED** [review]

File: docs/Telemetry.md [hunk 12]
```diff
@@ -215,6 +215,10 @@ Refer to the below table to understand different critical telemetry events raise
 | `ReceivedNullOrEmptyToken` | On receiving null or empty auth token |
 | `SystemMessageReceived` | On system message received |
 | `HistoryMessageReceived` | On history message received |
+| `SigninCardReceived` | On sign-in adaptive card received from bot |
+| `BotAuthActivityEmptySasUrl` | On detecting a signing adaptive card with no SAS Url |
```
Comment (charliewang95) [inline docs/Telemetry.md:219]: "signing" -> "sign-in"

File: docs/Telemetry.md (hunk 9 above)
Comment (charliewang95) [inline docs/Telemetry.md:220]: typo: "triger" -> "triggering"

**Pull request merged**